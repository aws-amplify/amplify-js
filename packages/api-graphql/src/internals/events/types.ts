// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Subscription } from 'rxjs';
import type { GraphQLAuthMode } from '@aws-amplify/core/internals/utils';

export interface SubscriptionObserver<T> {
	next(value: T): void;
	error(errorValue: any): void;
}

export interface EventsChannel {
	/**
	 * @experimental API may change in future versions
	 *
	 * Subscribe to Events
	 *
	 * @example
	 * const channel = await events.connect("default/channel")
	 *
	 * channel.subscribe({
	 *   next: (data) => { console.log(data) },
	 *   error: (err) => { console.error(err) },
	 * })
	 *
	 * @example // authMode override
	 * channel.subscribe({
	 *   next: (data) => { console.log(data) },
	 *   error: (err) => { console.error(err) },
	 * }, { authMode: 'userPool' })
	 *
	 * @param observer - observer callback handlers
	 * `{ next: () => {}, error: () => {}}`
	 *
	 * @param options - subscribe overrides: `authMode`, `authToken`
	 *
	 */
	subscribe(
		observer: SubscriptionObserver<any>,
		subOptions?: EventsOptions,
	): Subscription;
	/**
	 * @experimental API may change in future versions
	 *
	 * Close channel and any active subscriptions
	 *
	 * @example
	 * const channel = await events.connect("default/channel")
	 *
	 * channel.close()
	 *
	 */
	close(): void;
}

export type ResolvedGraphQLAuthModes = Exclude<GraphQLAuthMode, 'identityPool'>;

export interface EventsOptions {
	authMode?: GraphQLAuthMode;
	authToken?: string;
}

export interface PublishedEvent {
	identifier: string;
	index: number;
}

export interface PublishResponse {
	failed: PublishedEvent[];
	successful: PublishedEvent[];
}
