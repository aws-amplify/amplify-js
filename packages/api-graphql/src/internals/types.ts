// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { AmplifyClassV6 } from '@aws-amplify/core';
import { GraphQLAuthMode } from '@aws-amplify/core/internals/utils';
import { CustomHeaders } from '@aws-amplify/data-schema/runtime';

/**
 * @private
 *
 * The knobs available for configuring `generateClient` internally.
 */
export type ClientGenerationParams = {
	amplify: AmplifyClassV6;
} & CommonPublicClientOptions;

/**
 * Common options that can be used on public `generateClient()` interfaces.
 */
export interface CommonPublicClientOptions {
	authMode?: GraphQLAuthMode;
	authToken?: string;
	headers?: CustomHeaders;
}
