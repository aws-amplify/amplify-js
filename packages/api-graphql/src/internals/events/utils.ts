// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Amplify } from '@aws-amplify/core';
import {
	DocumentType,
	GraphQLAuthMode,
} from '@aws-amplify/core/internals/utils';

import type { ResolvedGraphQLAuthModes } from './types';

export const normalizeAuth = (
	explicitAuthMode: GraphQLAuthMode | undefined,
	defaultAuthMode: ResolvedGraphQLAuthModes,
): ResolvedGraphQLAuthModes => {
	if (!explicitAuthMode) {
		return defaultAuthMode;
	}

	if (explicitAuthMode === 'identityPool') {
		return 'iam';
	}

	return explicitAuthMode;
};

export const configure = () => {
	const config = Amplify.getConfig();

	const eventsConfig = config.API?.Events;

	if (!eventsConfig) {
		throw new Error(
			'Amplify configuration is missing. Have you called Amplify.configure()?',
		);
	}

	const configAuthMode = normalizeAuth(eventsConfig.defaultAuthMode, 'apiKey');

	const options = {
		appSyncGraphqlEndpoint: eventsConfig.endpoint,
		region: eventsConfig.region,
		authenticationType: configAuthMode,
		apiKey: eventsConfig.apiKey,
	};

	return options;
};

/**
 * Event API expects and array of JSON strings
 *
 * @param events - JSON-serializable value or an array of values
 * @returns array of JSON strings
 */
export const serializeEvents = (
	events: DocumentType | DocumentType[],
): string[] => {
	if (Array.isArray(events)) {
		return events.map((ev, idx) => {
			const eventJson = JSON.stringify(ev);
			if (eventJson === undefined) {
				throw new Error(
					`Event must be a valid JSON value. Received ${ev} at index ${idx}`,
				);
			}

			return eventJson;
		});
	}

	const eventJson = JSON.stringify(events);
	if (eventJson === undefined) {
		throw new Error(`Event must be a valid JSON value. Received ${events}`);
	}

	return [eventJson];
};
