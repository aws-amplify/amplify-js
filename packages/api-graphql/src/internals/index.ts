// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export {
	InternalGraphQLAPI,
	InternalGraphQLAPIClass,
} from './InternalGraphQLAPI';

export { graphql, cancel, isCancelError } from './v6';
export { generateClient } from './generateClient';
