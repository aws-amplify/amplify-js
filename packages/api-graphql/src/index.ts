// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as events from './internals/events';

export { events };

export { GraphQLAPI, GraphQLAPIClass, graphqlOperation } from './GraphQLAPI';
export * from './types';

export { CONNECTION_STATE_CHANGE } from './Providers/constants';
export * from './internals/events/types';
