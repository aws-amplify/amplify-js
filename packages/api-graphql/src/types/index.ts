// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { AmplifyClassV6, ResourcesConfig } from '@aws-amplify/core';
import {
	BaseClient,
	ClientExtensions,
	ClientExtensionsSSRCookies,
	ClientExtensionsSSRRequest,
	ClientInternals,
	CustomHeaders,
	ModelSortDirection,
} from '@aws-amplify/data-schema/runtime';
import { DocumentNode, GraphQLError, Source } from 'graphql';
import { Observable } from 'rxjs';
import {
	DocumentType,
	GraphQLAuthMode,
} from '@aws-amplify/core/internals/utils';
import { AmplifyServer } from '@aws-amplify/core/internals/adapter-core';

export { OperationTypeNode } from 'graphql';

export { CONTROL_MSG, ConnectionState } from './PubSub';

export { SelectionSet } from '@aws-amplify/data-schema/runtime';

export { CommonPublicClientOptions } from '../internals/types';

/**
 * Loose/Unknown options for raw GraphQLAPICategory `graphql()`.
 */
export interface GraphQLOptions {
	query: string | DocumentNode;
	variables?: Record<string, DocumentType>;
	authMode?: GraphQLAuthMode;
	authToken?: string;
	/**
	 * @deprecated This property should not be used
	 */
	userAgentSuffix?: string;
}

export interface GraphQLResult<T = object | null> {
	data: T;
	errors?: GraphQLError[];
	extensions?: Record<string, any>;
}

// Opaque type used for determining the graphql query type
declare const queryType: unique symbol;

export type GraphQLQuery<T> = T & { readonly [queryType]: 'query' };
export type GraphQLSubscription<T> = T & {
	readonly [queryType]: 'subscription';
};

/**
 * Accepts a code generated model type and returns a supertype that
 * can accept return values from the relevant graphql operations.
 *
 * For example:
 *
 * ```typescript
 * import { GraphQLReturnType } from 'aws-amplify/api';
 * import { MyModel } from './API';
 * import { getMyModel } from './graphql/queries';
 *
 * const [item, setItem] = useState<GraphQLReturnType<MyModel>>();
 * setItem(await client.graphql({ query: getMyModel }).data.getMyModel!)
 * ```
 *
 * Trying to assign the result of a `getMyModel` operation directly to a
 * `MyModel` variables won't necessarily work, because normally-required related
 * models might not be part of the selection set.
 *
 * This util simply makes related model properties optional recursively.
 */
export type GraphQLReturnType<T> =
	T extends Record<string, unknown>
		? {
				[K in keyof T]?: GraphQLReturnType<T[K]>;
			}
		: T;

/**
 * Describes a paged list result from AppSync, which can either
 * live at the top query or property (e.g., related model) level.
 */
interface PagedList<T, TYPENAME> {
	__typename: TYPENAME;
	nextToken?: string | null | undefined;
	items: T[];
}

/**
 * Recursively looks through a result type and removes nulls and
 * and undefined from `PagedList` types.
 *
 * Although a graphql response might contain empty values in an
 * array, this will only be the case when we also have errors,
 * which will then be *thrown*.
 */
type WithListsFixed<T> =
	T extends PagedList<infer IT, infer NAME>
		? PagedList<Exclude<IT, null | undefined>, NAME>
		: T extends Record<string, unknown>
			? {
					[K in keyof T]: WithListsFixed<T[K]>;
				}
			: T;

/**
 * Returns an updated response type to always return a value.
 */
type NeverEmpty<T> = {
	[K in keyof T]-?: Exclude<WithListsFixed<T[K]>, undefined | null>;
};

/**
 * Replaces all list result types in a query result with types to exclude
 * nulls and undefined from list member unions.
 *
 * If empty members are present, there will also be errors present,
 * and the response will instead be *thrown*.
 */
type FixedQueryResult<T> =
	Exclude<T[keyof T], null | undefined> extends PagedList<any, any>
		? {
				[K in keyof T]-?: WithListsFixed<Exclude<T[K], null | undefined>>;
			}
		: T;

/**
 * The return value from a `graphql({query})` call when `query` is a subscription.
 *
 * ```ts
 * //               |-- You are here
 * //               v
 * const subResult: GraphqlSubscriptionResult<T> = client.graphql({
 * 	query: onCreateWidget
 * });
 *
 * const sub = subResult.subscribe({
 * 	//
 * 	//            |-- You are here
 * 	//            v
 * 	next(message: GraphqlSubscriptionMessage<OnCreateWidgetSubscription>) {
 * 		handle(message.value);  // <-- type OnCreateWidgetSubscription
 * 	}
 * })
 * ```
 */
export type GraphqlSubscriptionResult<T> = Observable<
	GraphqlSubscriptionMessage<T>
>;

/**
 * The shape of messages passed to `next()` from a graphql subscription. E.g.,
 *
 * ```ts
 * const sub = client.graphql({
 * 	query: onCreateWidget,
 * }).subscribe({
 * 	//
 * 	//            |-- You are here
 * 	//            v
 * 	next(message: GraphqlSubscriptionMessage<OnCreateWidgetSubscription>) {
 * 		handle(message.value);  // <-- type OnCreateWidgetSubscription
 * 	}
 * })
 * ```
 */
export interface GraphqlSubscriptionMessage<T> {
	data: T;
}

export interface AWSAppSyncRealTimeProviderOptions {
	appSyncGraphqlEndpoint?: string;
	authenticationType?: GraphQLAuthMode;
	query?: string;
	variables?: Record<string, unknown>;
	apiKey?: string;
	region?: string;
	libraryConfigHeaders?(): () => Promise<Record<string, unknown> | Headers>;
	additionalHeaders?: CustomHeaders;
}

export interface AWSAppSyncRealTimeProvider {
	subscribe(
		options?: AWSAppSyncRealTimeProviderOptions,
	): Observable<Record<string, unknown>>;
}

export enum GraphQLAuthError {
	NO_API_KEY = 'No api-key configured',
	NO_CURRENT_USER = 'No current user',
	NO_CREDENTIALS = 'No credentials',
	NO_FEDERATED_JWT = 'No federated jwt',
	NO_AUTH_TOKEN = 'No auth token specified',
}

/**
 * GraphQLSource or string, the type of the parameter for calling graphql.parse
 * @see: https://graphql.org/graphql-js/language/#parse
 */
export type GraphQLOperation = Source | string;

/**
 * API V6 `graphql({options})` type that can leverage branded graphql `query`
 * objects and fallback types.
 */
export interface GraphQLOptionsV6<
	FALLBACK_TYPES = unknown,
	TYPED_GQL_STRING extends string = string,
> {
	query: TYPED_GQL_STRING | DocumentNode;
	variables?: GraphQLVariablesV6<FALLBACK_TYPES, TYPED_GQL_STRING>;
	authMode?: GraphQLAuthMode;
	authToken?: string;
	/**
	 * @deprecated This property should not be used
	 */
	userAgentSuffix?: string;
}

/**
 * Result type for `graphql()` operations that don't include any specific
 * type information. The result could be either a `Promise` or `Subscription`.
 *
 * Invoking code should either cast the result or use `?` and `!` operators to
 * navigate the result objects.
 */
export type UnknownGraphQLResponse =
	| Promise<GraphQLResult<any>>
	| GraphqlSubscriptionResult<any>;

/**
 * The expected type for `variables` in a V6 `graphql()` operation with
 * respect to the given `FALLBACK_TYPES` and `TYPED_GQL_STRING`.
 */
export type GraphQLVariablesV6<
	FALLBACK_TYPES = unknown,
	TYPED_GQL_STRING extends string = string,
> =
	TYPED_GQL_STRING extends GeneratedQuery<infer IN, any>
		? IN
		: TYPED_GQL_STRING extends GeneratedMutation<infer IN, any>
			? IN
			: TYPED_GQL_STRING extends GeneratedSubscription<infer IN, any>
				? IN
				: FALLBACK_TYPES extends GraphQLOperationType<infer IN, any>
					? IN
					: any;

/**
 * The expected return type with respect to the given `FALLBACK_TYPE`
 * and `TYPED_GQL_STRING`.
 */
export type GraphQLResponseV6<
	FALLBACK_TYPE = unknown,
	TYPED_GQL_STRING extends string = string,
> =
	TYPED_GQL_STRING extends GeneratedQuery<infer _, infer QUERY_OUT>
		? Promise<GraphQLResult<FixedQueryResult<QUERY_OUT>>>
		: TYPED_GQL_STRING extends GeneratedMutation<infer _, infer MUTATION_OUT>
			? Promise<GraphQLResult<NeverEmpty<MUTATION_OUT>>>
			: TYPED_GQL_STRING extends GeneratedSubscription<infer _, infer SUB_OUT>
				? GraphqlSubscriptionResult<NeverEmpty<SUB_OUT>>
				: FALLBACK_TYPE extends GraphQLQuery<infer _>
					? Promise<GraphQLResult<FALLBACK_TYPE>>
					: FALLBACK_TYPE extends GraphQLSubscription<infer _>
						? GraphqlSubscriptionResult<FALLBACK_TYPE>
						: FALLBACK_TYPE extends GraphQLOperationType<
									infer _,
									infer CUSTOM_OUT
							  >
							? CUSTOM_OUT
							: UnknownGraphQLResponse;

/**
 * The shape customers can use to provide `T` to `graphql<T>()` to specify both
 * `IN` and `OUT` types (the type of `variables` and the return type, respectively).
 *
 * I.E.,
 *
 * ```ts
 * type MyVariablesType = { ... };
 * type MyResultType = { ... };
 * type MyOperationType = { variables: MyVariablesType, result: MyResultType };
 *
 * const result: MyResultType = graphql<MyOperationType>("graphql string", {
 * 	variables: {
 * 		// MyVariablesType
 * 	}
 * })
 * ```
 */
export interface GraphQLOperationType<
	IN extends Record<string, DocumentType>,
	OUT extends Record<string, DocumentType>,
> {
	variables: IN;
	result: OUT;
}

/**
 * Nominal type for branding generated graphql query operation strings with
 * input and output types.
 *
 * E.g.,
 *
 * ```ts
 * export const getWidget = `...` as GeneratedQuery<
 * 	GetWidgetQueryVariables,
 * 	GetWidgetQuery
 * >;
 * ```
 *
 * This allows `graphql()` to extract `InputType` and `OutputType` to correctly
 * assign types to the `variables` and result objects.
 */
export type GeneratedQuery<InputType, OutputType> = string & {
	__generatedQueryInput: InputType;
	__generatedQueryOutput: OutputType;
};

/**
 * Nominal type for branding generated graphql mutation operation strings with
 * input and output types.
 *
 * E.g.,
 *
 * ```ts
 * export const createWidget = `...` as GeneratedQuery<
 * 	CreateWidgetMutationVariables,
 * 	CreateWidgetMutation
 * >;
 * ```
 *
 * This allows `graphql()` to extract `InputType` and `OutputType` to correctly
 * assign types to the `variables` and result objects.
 */
export type GeneratedMutation<InputType, OutputType> = string & {
	__generatedMutationInput: InputType;
	__generatedMutationOutput: OutputType;
};

/**
 * Nominal type for branding generated graphql mutation operation strings with
 * input and output types.
 *
 * E.g.,
 *
 * ```ts
 * export const createWidget = `...` as GeneratedMutation<
 * 	CreateWidgetMutationVariables,
 * 	CreateWidgetMutation
 * >;
 * ```
 *
 * This allows `graphql()` to extract `InputType` and `OutputType` to correctly
 * assign types to the `variables` and result objects.
 */
export type GeneratedSubscription<InputType, OutputType> = string & {
	__generatedSubscriptionInput: InputType;
	__generatedSubscriptionOutput: OutputType;
};

export const __amplify = Symbol('amplify');
export const __authMode = Symbol('authMode');
export const __authToken = Symbol('authToken');
export const __headers = Symbol('headers');

export function getInternals(client: BaseClient): ClientInternals {
	const c = client as any;

	return {
		amplify: c[__amplify],
		authMode: c[__authMode],
		authToken: c[__authToken],
		headers: c[__headers],
	} as any;
}

export type ClientWithModels =
	| V6Client
	| V6ClientSSRRequest
	| V6ClientSSRCookies;

export type V6Client<T extends Record<any, any> = never> = {
	graphql: GraphQLMethod;
	cancel(promise: Promise<any>, message?: string): boolean;
	isCancelError(error: any): boolean;
} & ClientExtensions<T>;

export type V6ClientSSRRequest<T extends Record<any, any> = never> = {
	graphql: GraphQLMethodSSR;
	cancel(promise: Promise<any>, message?: string): boolean;
	isCancelError(error: any): boolean;
} & ClientExtensionsSSRRequest<T>;

export type V6ClientSSRCookies<T extends Record<any, any> = never> = {
	graphql: GraphQLMethod;
	cancel(promise: Promise<any>, message?: string): boolean;
	isCancelError(error: any): boolean;
} & ClientExtensionsSSRCookies<T>;

export type GraphQLMethod = <
	FALLBACK_TYPES = unknown,
	TYPED_GQL_STRING extends string = string,
>(
	options: GraphQLOptionsV6<FALLBACK_TYPES, TYPED_GQL_STRING>,
	additionalHeaders?: CustomHeaders | undefined,
) => GraphQLResponseV6<FALLBACK_TYPES, TYPED_GQL_STRING>;

export type GraphQLMethodSSR = <
	FALLBACK_TYPES = unknown,
	TYPED_GQL_STRING extends string = string,
>(
	contextSpec: AmplifyServer.ContextSpec,
	options: GraphQLOptionsV6<FALLBACK_TYPES, TYPED_GQL_STRING>,
	additionalHeaders?: CustomHeaders | undefined,
) => GraphQLResponseV6<FALLBACK_TYPES, TYPED_GQL_STRING>;

/**
 * @private
 *
 * The knobs available for configuring `server/generateClient` internally.
 */
export interface ServerClientGenerationParams {
	amplify:
		| null // null expected when used with `generateServerClient`
		// closure expected with `generateServerClientUsingCookies`
		| ((fn: (amplify: AmplifyClassV6) => Promise<any>) => Promise<any>);
	// global env-sourced config use for retrieving modelIntro
	config: ResourcesConfig;
}

export type QueryArgs = Record<string, unknown>;

export interface ListArgs extends Record<string, unknown> {
	selectionSet?: string[];
	filter?: Record<string, unknown>;
	sortDirection?: ModelSortDirection;
	headers?: CustomHeaders;
}

export interface AuthModeParams extends Record<string, unknown> {
	authMode?: GraphQLAuthMode;
	authToken?: string;
}

export interface GenerateServerClientParams {
	config: ResourcesConfig;
	authMode?: GraphQLAuthMode;
	authToken?: string;
}
