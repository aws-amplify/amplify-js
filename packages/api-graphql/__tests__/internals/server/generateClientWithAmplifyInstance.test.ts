import { Amplify, AmplifyClassV6, ResourcesConfig } from '@aws-amplify/core';
import { generateClientWithAmplifyInstance } from '../../../src/internals/server';
import configFixture from '../../fixtures/modeled/amplifyconfiguration';
import { Schema } from '../../fixtures/modeled/schema';
import { V6ClientSSRRequest, V6ClientSSRCookies } from '../../../src/types';
import { mockApiResponse, normalizePostGraphqlCalls } from '../../utils';

const serverManagedFields = {
	id: 'some-id',
	owner: 'wirejobviously',
	createdAt: new Date().toISOString(),
	updatedAt: new Date().toISOString(),
};

const config: ResourcesConfig = {
	API: {
		GraphQL: {
			apiKey: 'apikey',
			customEndpoint: undefined,
			customEndpointRegion: undefined,
			defaultAuthMode: 'apiKey',
			endpoint: 'https://0.0.0.0/graphql',
			region: 'us-east-1',
		},
	},
};

describe('server generateClient', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	describe('with cookies', () => {
		test('subscriptions are disabled', () => {
			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			expect(() => {
				// @ts-expect-error
				client.models.Note.onCreate().subscribe();
			}).toThrow();
		});

		test('can list', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();

			expect(data.length).toBe(1);
			expect(data[0]).toEqual(
				expect.objectContaining({
					__typename: 'Todo',
					id: 'some-id',
					owner: 'wirejobviously',
					name: 'some name',
					description: 'something something',
				}),
			);
		});

		test('can list with nextToken', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
				nextToken: 'some-token',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						body: expect.objectContaining({
							// match nextToken in selection set
							query: expect.stringMatching(/^\s*nextToken\s*$/m),
						}),
					}),
				}),
			);
		});

		test('can list with limit', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
				limit: 5,
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						body: expect.objectContaining({
							// match nextToken in selection set
							query: expect.stringMatching(/^\s*nextToken\s*$/m),
						}),
					}),
				}),
			);
		});

		test('can list with sort direction (ascending)', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				id: 'some-id',
				sortDirection: 'ASC',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
		});
		test('can list with sort direction (descending)', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				id: 'some-id',
				sortDirection: 'DESC',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
		});

		test('can custom query', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					echo: {
						resultContent: 'echo result content',
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const result = await client.queries.echo({
				argumentContent: 'echo argumentContent value',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
			expect(result?.data).toEqual({
				resultContent: 'echo result content',
			});
		});
		describe('with request', () => {
			test('subscriptions are disabled', () => {
				const client = generateClientWithAmplifyInstance<
					Schema,
					V6ClientSSRRequest<Schema>
				>({
					amplify: null,
					config: config,
				});

				expect(() => {
					// @ts-expect-error
					client.models.Note.onCreate().subscribe();
				}).toThrow();
			});

			test('contextSpec param gets passed through to client.graphql', async () => {
				Amplify.configure(configFixture as any);
				const config = Amplify.getConfig();

				const client = generateClientWithAmplifyInstance<
					Schema,
					V6ClientSSRRequest<Schema>
				>({
					amplify: null,
					config: config,
				});

				const mockContextSpec = {};

				const spy = jest
					.spyOn(client, 'graphql')
					.mockImplementation(async () => {
						const result: any = {};
						return result;
					});

				await client.models.Note.list(mockContextSpec);

				expect(spy).toHaveBeenCalledWith(
					mockContextSpec,
					expect.objectContaining({
						query: expect.stringContaining('listNotes'),
					}),
					{},
				);
			});

			test('can custom query', async () => {
				Amplify.configure(configFixture as any);
				const config = Amplify.getConfig();

				const client = generateClientWithAmplifyInstance<
					Schema,
					V6ClientSSRRequest<Schema>
				>({
					amplify: null,
					config: config,
				});

				const spy = jest
					.spyOn(client, 'graphql')
					.mockImplementation(async () => {
						const result: any = {};
						return result;
					});

				const mockContextSpec = { token: { value: Symbol('test') } };

				const result = await client.queries.echo(mockContextSpec, {
					argumentContent: 'echo argumentContent value',
				});

				expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
			});
		});
	});
});
