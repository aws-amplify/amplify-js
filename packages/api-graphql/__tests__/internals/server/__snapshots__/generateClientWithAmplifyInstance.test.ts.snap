// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server generateClient with cookies can custom query 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($argumentContent: String!) {
  echo(argumentContent: $argumentContent) {
    resultContent
  }
}
",
          "variables": {
            "argumentContent": "echo argumentContent value",
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies can list 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($filter: ModelTodoFilterInput, $limit: Int, $nextToken: String) {
  listTodos(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      description
      status
      tags
      createdAt
      updatedAt
      todoMetaId
      owner
    }
    nextToken
    __typename
  }
}
",
          "variables": {
            "filter": {
              "name": {
                "contains": "name",
              },
            },
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies can list with limit 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($filter: ModelTodoFilterInput, $limit: Int, $nextToken: String) {
  listTodos(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      description
      status
      tags
      createdAt
      updatedAt
      todoMetaId
      owner
    }
    nextToken
    __typename
  }
}
",
          "variables": {
            "filter": {
              "name": {
                "contains": "name",
              },
            },
            "limit": 5,
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies can list with nextToken 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($filter: ModelTodoFilterInput, $limit: Int, $nextToken: String) {
  listTodos(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      description
      status
      tags
      createdAt
      updatedAt
      todoMetaId
      owner
    }
    nextToken
    __typename
  }
}
",
          "variables": {
            "filter": {
              "name": {
                "contains": "name",
              },
            },
            "nextToken": "some-token",
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies can list with sort direction (ascending) 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($filter: ModelTodoFilterInput, $limit: Int, $nextToken: String) {
  listTodos(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      description
      status
      tags
      createdAt
      updatedAt
      todoMetaId
      owner
    }
    nextToken
    __typename
  }
}
",
          "variables": {
            "id": "some-id",
            "sortDirection": "ASC",
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies can list with sort direction (descending) 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($filter: ModelTodoFilterInput, $limit: Int, $nextToken: String) {
  listTodos(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      description
      status
      tags
      createdAt
      updatedAt
      todoMetaId
      owner
    }
    nextToken
    __typename
  }
}
",
          "variables": {
            "id": "some-id",
            "sortDirection": "DESC",
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with cookies with request can custom query 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($argumentContent: String!) {
      echo(argumentContent: $argumentContent) {resultContent}
    }
  ",
      "variables": {
        "argumentContent": "echo argumentContent value",
      },
    },
  ],
]
`;
